src/game/AI/ScriptDevAI/scripts/northrend/naxxramas/boss_sapphiron.cpp

```diff
@@ -16,339 +16,398 @@
                     {
-                        DoCastSpellIfCan(m_creature, SPELL_FROST_BREATH_A, CAST_TRIGGERED);
-                        DoCastSpellIfCan(m_creature, SPELL_FROST_BREATH_B, CAST_TRIGGERED);
-                        m_uiFrostBreathTimer = 0;
-
-                        m_uiLandTimer = 4000;
+                        ++m_iceboltCount;
+                        ResetCombatAction(action, GetSubsequentActionTimer(action));
                     }
-                    else
-                        m_uiFrostBreathTimer -= uiDiff;
                 }
-
-                if (m_uiLandTimer)
+                else
                 {
-                    if (m_uiLandTimer <= uiDiff)
-                    {
-                        // Begin Landing
-                        DoScriptText(EMOTE_GROUND, m_creature);
-                        m_creature->HandleEmote(EMOTE_ONESHOT_LAND);
-                        m_creature->SetLevitate(false);
-
-                        m_Phase = PHASE_LANDING;
-                        m_uiLandTimer = 2000;
-                    }
-                    else
-                        m_uiLandTimer -= uiDiff;
+                    m_iceboltCount = 0;
+                    DisableCombatAction(action);
+                    ResetCombatAction(SAPPHIRON_FROST_BREATH, 100u);    // Enough Icebolts were cast, switch to Frost Breath (Ice Bomb) after that
                 }
-
-                break;
-            case PHASE_LANDING:
-                if (m_uiLandTimer < uiDiff)
+                return;
+            }
+            case SAPPHIRON_FROST_BREATH:
+            {
+                if (DoCastSpellIfCan(m_creature, SPELL_FROST_BREATH) == CAST_OK)
                 {
-                    m_Phase = PHASE_GROUND;
-                    SetDeathPrevention(false);
-                    SetCombatMovement(true);
-                    m_creature->GetMotionMaster()->Clear(false);
-                    m_creature->GetMotionMaster()->MoveChase(m_creature->GetVictim());
-
-                    m_uiFlyTimer = 67000;
-                    m_uiLandTimer = 0;
+                    DoCastSpellIfCan(m_creature, SPELL_FROST_BREATH_DUMMY, CAST_TRIGGERED);
+                    DoScriptText(EMOTE_BREATH, m_creature);
+                    DisableCombatAction(action);
+                    ResetTimer(SAPPHIRON_LANDING_PHASE, 10u * IN_MILLISECONDS);
                 }
-                else
-                    m_uiLandTimer -= uiDiff;
-
-                break;
-        }
-
-        // Enrage can happen in any phase
-        if (m_uiBerserkTimer < uiDiff)
-        {
-            if (DoCastSpellIfCan(m_creature, SPELL_BESERK) == CAST_OK)
+                return;
+            }
+            case SAPPHIRON_BERSERK:
             {
-                DoScriptText(EMOTE_GENERIC_ENRAGED, m_creature);
-                m_uiBerserkTimer = 300000;
+                if (DoCastSpellIfCan(m_creature, SPELL_BESERK) == CAST_OK)
```

Comment:
Berserk typo, also spell is TARGET_UNIT_CASTER only, as such, m_creature should not be passed as target